File: jiseok/LeetCode/3101.py
class Solution:
    def countAlternatingSubarrays(self, nums: List[int]) -> int:
        # O(N)이나 O(NlogN)
        # 리스트에다가 지금까지 몇 개의 수가 달랐는가를 저장
        # 그럼 리스트가 의미하는 게 해당 인덱스까지의 가능한 subarray 개수 의미
        # 그냥 count하면 끝?
        N = len(nums)
        # cnt_list = [1 for _ in range(N)]
        # for i in range(1, N):
        #     if nums[i] != nums[i - 1]:
        #         cnt_list[i] = cnt_list[i - 1] + 1

        # return sum(cnt_list)


        # Runtime / Memory 맘에 안 들어서 리스트 안 쓰고 풀이
        ans = 1
        cur_acc = 1
        for i in range(1, N):
            if nums[i] != nums[i - 1]:
                cur_acc += 1
            else:
                cur_acc = 1
            ans += cur_acc

        return ans

File: dong99u/leetcode/3101.java
package leetcode.leetcode3101;

/**
 * 1 <= n <= 10^5 이므로
 * brute force로 풀면 O(n^2) or O(n^3)이라 시간초과 남.
 * 적어도 O(nlogn)이하로 풀기 -> dp?
 */
public class Solution {

	public static void main(String[] args) {
		System.out.println(countAlternatingSubarrays(new int[] {1, 0, 1, 0}));
	}

	public static long countAlternatingSubarrays(int[] nums) {
		int n = nums.length;

		int[] dp = new int[n]; // i번째 인덱스로 끝나는 교대배열의 개수
		dp[0] = 1;

		/**
		 * nums = [0, 1, 0, 1]
		 * dp =>  0  1  0   1
		 *          01 10  01
		 *             010 101
		 *                 0101
		 */
		for (int i = 1; i < n; i++) {
			if (nums[i] != nums[i - 1]) {
				dp[i] = dp[i - 1] + 1;
			} else {
				dp[i] = 1;
			}
		}

		long sum = 0;
		for (int elem : dp) {
			sum += elem;
		}

		return sum;
    }
}
